### PR TITLE
`ASAP-Alchemy`: Inject experimental ligands in to networks

### LINKED ISSUE #897: `alchemy`: automatically add refernce compounds to a network
To improve the accuracy of our free energy predictions we should be adding reference compounds to our planned networks which have assay data in CDD. 

MVP:
The plan is to extend the `prep` stage to accept a CDD protocol name and the number of compounds which we should try and add. Then after sorting by similarity to the crystal reference compound we should try and generate a pose for each molecule until either we hit the requested number or run out of molecules. These compounds will then be tagged as experimental reference compounds. 

Stretch:
Penalise the network planner to not add connections between the experimental compounds, this should mean that the compounds are distributed through the network and then during the predict stage we can add extra edges with the experimentally determined differences which should help improve the accuracy of the network.

cc: @JenkeScheen

### PR DESCRIPTION
**Pull request by jthorton**
## Description
This PR extends the `ASAP-Alchemy` prep workflow to automatically add experimental ligands into a network from the provided CDD experimental protocol.

closes #897 

## Workflow:
- Download all ligands with valid experimental data in the given CCD protocol
- Extract all ligands with fully defined stereochemistry
- Sort the ligands by MCS overlap with the crystal ligand (to ensure a valid pose can be generated)
- Remove any experimental ligands which are in the target list (ensure no self-transforms)
- Iteratively generate poses for the experimental ligands until we have the user-provided N (default =3) experimental ligands posed in the dataset
- Mark the ligands as experimental to ensure experimental edges are added during the prediction stage to strengthen the network. 

## Todos
Notable points that this PR has either accomplished or will accomplish.
- [x] Inject experimental ligands
- [ ] identify experimental ligands during prediction and add edges, this is now being moved to another PR
- [x] Add tests of full workflow

## Questions
- [ ] Question 1 ..

## Status
- [x] Ready to go

## Developers certificate of origin
- [x] I certify that this contribution is covered by the MIT license as defined in our [LICENSE](https://github.com/choderalab/asapdiscovery/blob/main/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

### CODE REVIEW

**Comment by JenkeScheen** [comment]
testing this in production for a prospective project support ticket today

**Comment by JenkeScheen** [comment]
this seems to be working as intended, awesome! One change I'd suggest is changing the order to:
1. find the `n` compounds to inject
2. add the `n` compounds to the `N` compounds in the moleculeset
3. do pose generation

Because currently it looks like it's:
1. find the `n` compounds to inject
2. generate poses for `N` compounds in moleculeset
3. generate poses for `n` injected compounds
4. add `n` poses to `N` poses

**Comment by JenkeScheen** [comment]
crap - I accidentally committed to this branch thinking I'd made it for finding crystal references, whoops. Signing off now, will fix tomorrow

**Comment by JenkeScheen** [comment]
@jthorton I believe I've now reverted all my commits without affecting yours - please double-check : )

**Comment by jthorton** [comment]
@JenkeScheen thanks for fixing this up! 
> this seems to be working as intended, awesome! One change I'd suggest is changing the order to:

    find the n compounds to inject
    add the n compounds to the N compounds in the moleculeset
    do pose generation
    Because currently it looks like it's:
    
    find the n compounds to inject
    generate poses for N compounds in moleculeset
    generate poses for n injected compounds
    add n poses to N poses
    
This was the original plan to streamline things but I noticed I would often get fewer ligands than requested as we can not tell if the stereochemistry of the core matches the reference ligand until we pose it. So I opted to iteratively pose and check each molecule and end the loop once we have the required number we can look at changing this up in future though!

**Comment by jthorton** [comment]
@hmacdope any idea why the workflows CI just gets cancelled?

**Comment by hmacdope** [comment]
Very odd, ill look

File: asapdiscovery-data/asapdiscovery/data/operators/selectors/mcs_selector.py [hunk 1]
```diff
@@ -13,6 +13,88 @@
 logger = logging.getLogger(__name__)
 
 
+def sort_by_mcs(
```
Comment (jthorton) [inline asapdiscovery-data/asapdiscovery/data/operators/selectors/mcs_selector.py:16]: @hmacdope the failing tests could be due to changes here I think this is the only thing related to the workflow tests?

**Comment by jthorton** [comment]
So workflows passes now, it must be something todo with how I changed the mcs matcher?

File: asapdiscovery-data/asapdiscovery/data/operators/selectors/mcs_selector.py (hunk 1 above)
Comment (jthorton) [reply asapdiscovery-data/asapdiscovery/data/operators/selectors/mcs_selector.py:16]: reversing the change fixed the tests on CI.  I tried to reproduce the long test time locally but failed so not sure what is happening. I'll try again tomorrow.

**Comment by hmacdope** [comment]
Lets just go ahead here.

**Pull request merged**